### DISCUSSION TITLE
Add NPM Audit data to add more detail to Renovate PRs or limit to merge requests to only vulnerability fixes

### DISCUSSION THREAD

**Discussion by fullstackzach** [renovatebot/renovate#22176] [Suggest an Idea]
**What would you like Renovate to be able to do?**
Renovate upgrades packages to the latest that is available, but I would like bring in `NPM Audit` data

**Describe the solution you'd like**

It could possibly:
1. limit/filter Renovate to only PR versions reported as per data "fixed by" NPM Audit, so that pull requests would only be opened for modules that have vulnerability issues.
2. and/or add useful information to a PR that shows the count and level of severity of vulnerabilities fixed by a renovate pull request.
3. Ability to specify a `severity-level` similar to npm audit to target PRs for.

**Describe alternatives you've considered**
Some implementation ideas:
  - Run `npm audit --json` on each repo to get audit fix data
  - Use npm audit's rest API directly to get data on dependencies (Interesting example here: https://dzone.com/articles/how-to-use-npm-rest-api-to-get-audit-npm-audit-res)

**Additional context**

This seems Related to #3080, but it talks about GitHub vulnerability alerts.  For my usecase I would be looking for this on a self-hosted instance of GitLab using the renovate npm module. 

I'd be interested in digging in and helping make a contribution for this.

**Comment by rarkins** [comment]
This could be very useful but definitely needs a decent amount of planning before I recommend starting any *implementation*. I've been mostly guilty myself of hacking in features in the past that now collectively often make it really hard to simplify/refactor code when I want to.

e.g. when is the audit data retrieved:
 - at the start of the run like we do with GitHub vulnerability alerts? Kind of hard because we don't yet know where the package files all are
 - post-extraction, once we know what's in all the package files? probably
 - at the time when we normally call `npm`? I think this would be too late because we're already half way through making the PR

**Comment by fullstackzach** [comment]
Thanks @rarkins - I'll do some analysis and come up with some ideas

**Comment by pahlers** [comment]
@rarkins Please, can I bring this feature to your attention again? 

It would be a very helpful tool to fight against supply-chain attacks. There is a thread called [Use compromised dependency](https://slsa.dev/spec/v1.0/threats-overview) and this feature would help to lower the risks.

**Another tools beside npm-audit**

There is also a product called [OWASP dependency-check](https://owasp.org/www-project-dependency-check/) which supports multiple analyzers.
- Archive
- Assembly
- Jar
- RetireJS
- Node.js
- Node Audit
- Nugetconf
- Nuspec
- OpenSSL
- OSS
- Ruby bundler‑audit

With multiple sources
- [NATIONAL VULNERABILITY DATABASE](https://nvd.nist.gov/)
- [CISA Catalog of Known Exploited Vulnerabilities](https://www.cisa.gov/known-exploited-vulnerabilities-catalog)
- [OSS Index](https://ossindex.sonatype.org/)

A few downsides
- its slow, depending on the caching it can be from 2 up to 60 minutes when an analyse is completed.
- its written in Java, which isn't Renovates first choice.

**Reply by rarkins** [reply]
See https://docs.renovatebot.com/configuration-options/#osvvulnerabilityalerts

**Reply by pahlers** [reply]
That is a fast reaction. Thanks!

> You will only get OSV-based vulnerability alerts for direct dependencies. Renovate only queries the OSV database for dependencies that use one of these datasources:...

Which means "direct dependencies". Are those only the dependencies mentioned in, for example, the package.json, or also recursive down the dependency tree (as in the package-lock.json)?

**Reply by rarkins** [reply]
package.json, pom.xml, etc. Not lock files.

**Reply by pahlers** [reply]
That is unfortunate. Than this feature won't help us against the `Use compromised dependency` thread.

I will mention this in https://github.com/renovatebot/renovate/discussions/20542, because it's very important. I hope we can pickup the discussion there.

**Reply by rarkins** [reply]
Are you willing to contribute such features in PRs, or just hoping to motivate others to do so?

Ideas are welcome, contributions even more welcome, but "hey someone else should do this" comments should be kept to a minimum.

**Reply by pahlers** [reply]
I hope I can contribute. I already took a look at the source code. I didn't grasped the structure at ones, but it looked promising. 
The reason I looked to your code was that I want to build a periodic dependency-check my self and I looked into your code to see how you did this. And I thought, how cool would it be when Renovate can do this?

I have some experience with build platforms, static code analysis and preventing supply chain risks.

As you said before there is some planning involved. Probably you have already some idea's how you want to solve this.

I will start by reading some development documents, if there are. Else I will look into  the code. Can you recommend me some documents?

**Reply by HonkingGoose** [reply]
## Describe vulnerability management tools in Renovate docs?

`@rarkins` How about putting the audit tools info, and maybe more, in the Renovate docs proper? That way we can help our users keep themselves safe a bit more. We would need to check the recommendations we give are:

- well known
- well maintained
- safe to use (as far as we know)

### Where to put the new info?

I don't know yet what's best, a new Deep Dive article:

1. New Deep Dive article called _Vulnerability check tools_
2. Or, add info to _Upgrade best practices_ docs.

### What's the scope?

Do we want to point out audit tools _only_, or also related things like linters, static code checkers and so on?

#### Audit tools

- `npm audit`
- OWASP dependency check

#### Linters

- Shellcheck (to prevent unsafe code in Bash scripts)
- ESLint
- Other relevant static code checkers

### Tell me what you think!

As usual, let me know if this is something you as maintainer want to have, and how it should look in broad terms. :wink: 

## Development docs

> I will start by reading some development documents, if there are. Else I will look into the code. Can you recommend me some documents?

@pahlers please read the [Renovate docs, Contributing to Renovate](https://docs.renovatebot.com/contributing-to-renovate/), especially the _I want to write code_ bits.
